### PR TITLE
Add unit test for get_scaled_labelled_x_star

### LINKED ISSUE #29: Add unit test: matrix_a_star: `get_scaled_labelled_x_star`
The function `get_scaled_labelled_x_star` does not have a unit test. It should have one.

### PR DESCRIPTION
**Pull request by danasebo**
## Description
Add unit test for get_scaled_labelled_x_star. This involved moving some stuff to fixtures stored in the module conftest.py script.

Closes [#29](https://github.com/ONSdigital/scalelink/issues/29)

## Type of change

*You can delete options that are not relevant.*

- [ ] Bug fix - *non-breaking change*
- [x] New feature - *non-breaking change*
- [ ] Breaking change - *backwards incompatible change, changes expected behaviour*
- [x] Non-user facing change, structural change, dev functionality, docs ...

## Checklist:

- [x] I have performed a self-review of my own code.
- [x] I have commented my code appropriately, focusing on explaining my design decisions (explain why, not how).
- [x] I have made corresponding changes to the documentation (comments, docstring, etc.. )
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have updated the change log.

<br>

#  Peer review
Any new code includes all the following:

- **Documentation**: docstrings, comments have been added/ updated.
- **Style guidelines**: New code conforms to the project's contribution guidelines.
- **Functionality**: The code works as expected, handles expected edge cases and exceptions are handled appropriately.
- **Complexity**: The code is not overly complex, logic has been split into appropriately sized functions, etc..
- **Test coverage**: Unit tests cover essential functions for a reasonable range of inputs and conditions. Added and existing tests pass on my machine.

### Review comments
Suggestions should be tailored to the code that you are reviewing. Provide context.
Be critical and clear, but not mean. Ask questions and set actions.
<details><summary>These might include:</summary>

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented
  - Do the tests effectively assure that it
  works correctly? Are there additional edge cases/ negative tests to be considered?
- code style improvements (could the code be written more clearly?)
</details>
<br>

*Further reading: [code review best practices](https://best-practice-and-impact.github.io/qa-of-code-guidance/peer_review.html)*

### CODE REVIEW

File: tests/matrix_a_star/conftest.py [hunk 1]
```diff
@@ -56,3 +58,33 @@ def make_matrix_a_output() -> np.array:
...
     )
+
+
+@pytest.fixture(scope="module")
+def make_input_matrix_a_star() -> np.array:
+    """
+    Sets up the input matrix_a_star for test_solve_for_x_star, and it is also
+    one of the test inputs for test_get_scaled_labelled_x_star.
+    """
+
```
Comment (mary-cleaton) [inline tests/matrix_a_star/conftest.py]: Here and below you can probably do without the line break between the docstring and the return value. The formatter might put it back (not sure!) but personally I tend not to put a line break after a function docstring.

File: tests/matrix_a_star/test_matrix_a_star.py [hunk 2]
```diff
@@ -176,9 +176,60 @@ def test_get_matrix_a_star(
     )
 
 
-@pytest.mark.skip(reason="test shell")
-def test_get_scaled_labelled_x_star() -> None:
-    pass
+@patch("scalelink.matrix_a_star.matrix_a_star.scale_x_star")
```
Comment (mary-cleaton) [inline tests/matrix_a_star/test_matrix_a_star.py:179]: Confess I've only skim-read this. However, it looks exactly right.

I also ran all tests locally and all passed.

**Review by mary-cleaton — APPROVED** [review]
All looks good to me - well done!

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 2 above)
Comment (grthomson) [reply tests/matrix_a_star/test_matrix_a_star.py:179]: Have also run this locally with pytest.  The assertions all pass and confirm that all mocked dependencies called as expected.

File: tests/matrix_a_star/test_matrix_a_star.py [hunk 3]
```diff
@@ -176,9 +176,60 @@ def test_get_matrix_a_star(
...
+        "cutpoints": {"sex": None, "fn": [0.75, 0.9]},
+        "K": 3,
+        "s": 7,
+    }
+    test_b_output = [0, 0, 0, 0, 1]
+    test_label_x_star_output = {
+        "sex_disagree": 2,
+        "sex_agree": 1,
+    }
+
```
Comment (grthomson) [inline tests/matrix_a_star/test_matrix_a_star.py:213]: I think that it is still worthwhile to define `test_scale_x_star_output`.

Currently the final assert condition checks that `mock_scale_x_star` is called on the output of `label_x_star`.

If you include e.g.:

```
test_scale_x_star_output = {
        "sex_disagree": 0.0,  
        "sex_agree" : -100.0  
 }
```

then you can use an assert to check that the return value of `get_scaled_labelled_x_star` just is the unchanged return value of `scale_x_star`. This is just a final check that the function is wired correctly.

File: tests/matrix_a_star/test_matrix_a_star.py [hunk 4]
```diff
@@ -176,9 +176,60 @@ def test_get_matrix_a_star(
...
+    mock_calculate_b.return_value = test_b_output
+    mock_multiply_vectors_by_s.return_value = make_input_b
+    mock_solve_for_x_star.return_value = make_solve_for_x_star_output
+    mock_label_x_star.return_value = test_label_x_star_output
+
```
Comment (grthomson) [inline tests/matrix_a_star/test_matrix_a_star.py:219]: If using `test_scale_x_star_output` the following lines would be included:

```
mock_scale_x_star.return_value = test_scale_x_star_output

# Act
result = ma.get_scaled_labelled_x_star(
    matrix_a_star=make_input_matrix_a_star, input_variables=test_input_dict
 )
```

File: tests/matrix_a_star/test_matrix_a_star.py [hunk 5]
```diff
@@ -176,9 +176,60 @@ def test_get_matrix_a_star(
...
+        vector=test_b_output,
+        s=test_input_dict["s"],
+    )
+    mock_solve_for_x_star.assert_called_once_with(
+        matrix_a_star=make_input_matrix_a_star, b=make_input_b
+    )
+    mock_label_x_star.assert_called_once_with(
+        x_star=make_solve_for_x_star_output, cutpoints=test_input_dict["cutpoints"]
+    )
+    mock_scale_x_star.assert_called_once_with(x_star_labelled=test_label_x_star_output)
```
Comment (grthomson) [inline tests/matrix_a_star/test_matrix_a_star.py:239]: Final assert condition:

```assert result == mock_scale_x_star.return_value```

File: tests/matrix_a_star/conftest.py (hunk 1 above)
Comment (danasebo) [reply tests/matrix_a_star/conftest.py]: I agree, thank you! Fixed now, so it is consistent will all other functions.

Comment (mary-cleaton) [reply tests/matrix_a_star/conftest.py]: Agree this is fixed. Thanks!

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 3 above)
Comment (mary-cleaton) [reply tests/matrix_a_star/test_matrix_a_star.py:213]: I like this suggestion, although it would require this step to be added to all of the tests of parent functions.

@danasebo - can you please:

1. Do this for `test_get_scaled_labelled_x_star` in this branch.
2. Raise an issue for doing it to all the other parent functions. You can easily detect the parent functions by looking at the module docstrings in the files within the subfolders of the `scalelink` folder - the parent functions have an entry in the module docstring that describes which child functions they call (and in which order).
3. If you like and have time, pick up and complete that issue.

Comment (danasebo) [reply tests/matrix_a_star/test_matrix_a_star.py:213]: Thanks Gavin, this suggestion has been implemented.
I will raise an issue for doing it to all the other parent functions.

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 4 above)
Comment (danasebo) [reply tests/matrix_a_star/test_matrix_a_star.py:219]: thank you for your suggestion and helpful guidance. done :)

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 5 above)
Comment (danasebo) [reply tests/matrix_a_star/test_matrix_a_star.py:239]: done.

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 3 above)
Comment (grthomson) [reply tests/matrix_a_star/test_matrix_a_star.py:213]: Have run updated tests locally with `pytest` and all looking good.

**Review by grthomson — APPROVED** [review]
Unit test and new assert condition tested locally with `pytest`, looks good and all passing.
